### PR TITLE
[KFKESQ-12] Allow the configuration properties of consumer/producer 

### LINKED ISSUE #19: [KFKESQ-12] Allow the configuration of internal KafkaConsumer and KafkaProducer using system properties
Allow the configuration of internal `KafkaConsumer` and `KafkaProducer` using system properties.

As an example, we can try to implement the same approach concerning the properties used by the `spring-kafka` library (see [here](https://docs.spring.io/spring-boot/docs/current/reference/html/application-properties.html#application-properties.integration.spring.kafka.consumer.properties)).

The `KafkaConsumer` / `KafkaProducer` properties should be rooted at some paths like `kafkaesque.consumer.properties.*` and `kafkaesque.producer.properties.*`.

### PR DESCRIPTION
**Pull request by farajist**
Kafka's internal consumer and producer can now be configured using a `kafkaesque.properties` file in [spring-kafka](https://docs.spring.io/spring-boot/docs/current/reference/html/application-properties.html#appendix.application-properties.integration)'s fashion, this has been implemented with extensibility in mind, and [OWNER](https://matteobaccan.github.io/owner/) seems to fit the bill quite elegantly. Please check and let me know what you think  @rcardin

P.S: it's been a busy week but I finally got to sit down and work on it sorry for the delay :sweat_smile:

Fixes #19

### CODE REVIEW

**Comment by rcardin** [comment]
Hey, man. Quite a busy week here too. I plan to review your PR during the weekend.

File: kafkaesque-core/src/main/java/in/rcard/kafkaesque/consumer/KafkaesqueConsumer.java [hunk 1]
```diff
@@ -57,17 +58,20 @@
...
-    props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
-    props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, brokersUrl);
-    props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "true");
-    props.put(
-        ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, creationInfo.getKeyDeserializer().getClass());
-    props.put(
-        ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
-        creationInfo.getValueDeserializer().getClass());
-    final KafkaConsumer<Key, Value> consumer = new KafkaConsumer<>(props);
+    ConfigFactory.setProperty("configFile", "kafkaesque");
```
Comment (rcardin) [inline kafkaesque-core/src/main/java/in/rcard/kafkaesque/consumer/KafkaesqueConsumer.java:61]: I prefer to leave the implementation of the config mechanism behind some interface. In this way, we are less coupled with external dependencies.

The comment also applies to the types `KafkaesqueConsumerConfig`, `KafkaesqueProducerConfig`, and so on.

Comment (rcardin) [inline kafkaesque-core/src/main/java/in/rcard/kafkaesque/consumer/KafkaesqueConsumer.java:61]: Can't we set this property of the `ConfigFactory` away from here, i.e., in a place where it's set only once?

File: kafkaesque-core/src/main/java/in/rcard/kafkaesque/consumer/KafkaesqueConsumerConfigMapper.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package in.rcard.kafkaesque.consumer;
+
+import java.util.Map;
+import java.util.Properties;
+
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+
+public class KafkaesqueConsumerConfigMapper {
+
+    public static Map toConsumerProperties(KafkaesqueConsumerConfig kConsumerConfig) {
```
Comment (rcardin) [inline kafkaesque-core/src/main/java/in/rcard/kafkaesque/consumer/KafkaesqueConsumerConfigMapper.java:10]: Don't use raw types. Let the compiler help us 😄

File: kafkaesque-core/src/main/java/in/rcard/kafkaesque/consumer/KafkaesqueConsumer.java [hunk 3]
```diff
@@ -57,17 +58,20 @@
...
+            Map.of(
+                    KAFKAESQUE_CONSUMER_CONFIG_BOOTSTRAP_SERVERS,
+                    brokersUrl,
+                    KAFKAESQUE_CONSUMER_CONFIG_KEY_DESERIALIZER,
+                    creationInfo.getKeyDeserializer().getClass().getName(),
+                    KAFKAESQUE_CONSUMER_CONFIG_VALUE_DESERIALIZER,
+                    creationInfo.getValueDeserializer().getClass().getName()
+            )
+    );
+    final Map consumerProperties = KafkaesqueConsumerConfigMapper.toConsumerProperties(consumerConfig);
```
Comment (rcardin) [inline kafkaesque-core/src/main/java/in/rcard/kafkaesque/consumer/KafkaesqueConsumer.java:73]: Don't use raw types.

File: kafkaesque-core/src/main/java/in/rcard/kafkaesque/consumer/KafkaesqueConsumerConfigMapper.java (hunk 2 above)
Comment (rcardin) [inline kafkaesque-core/src/main/java/in/rcard/kafkaesque/consumer/KafkaesqueConsumerConfigMapper.java:10]: You don't need the `KafkaesqueConsumerConfigMapper` type at all. You can move the method `toConsumerProperties` directly in the `KafkaesqueConsumerConfig` class.

**Review by rcardin — CHANGES_REQUESTED** [review]
Hey, thanks for the PR. Nice work. I left some comments. IMO, we must decouple the implementation we choose to read configuration from the rest of the Kafkaesque code.

The comments I left for the consumer part also apply to the producers.

Feel free to let me know if you need more details.

**Comment by farajist** [comment]
Hey @rcardin, I've worked on a second iteration of the PR based on the code review comments, I didn't pay much attention to the license issue at first (since I'm not too familiar with how it works), but I guess the alternative lib used now is compatible with MIT license, other than that the refactored code might seem a little more verbose, please check and let me know what you are thinking :smile:

**Comment by rcardin** [comment]
Hey, man. Sorry for the delay; crazy times at work. I'll try to review your last commits this week.

File: README.md [hunk 4]
```diff
@@ -115,3 +115,22 @@ Since the library is published in the GitHub Packages repository, you need to ad
   <url>https://maven.pkg.github.com/rcardin/kafkaesque</url>
 </repository>
 ```
+
+#### Configuration
+
+_Kafkaesque_ also supports internal producer and consumer configuration via an `application.properties`, `application.conf`  or `application.json` configuration file.
```
Comment (rcardin) [inline README.md:121]: What about YAML files? Where should the file be placed?

File: kafkaesque-core/src/main/java/in/rcard/kafkaesque/config/KafkaesqueConfigLoader.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package in.rcard.kafkaesque.config;
+
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigBeanFactory;
+import com.typesafe.config.ConfigFactory;
+
+public class KafkaesqueConfigLoader {
+
+    public static KafkaesqueConsumerConfig loadConsumerConfig() {
+        Config config = ConfigFactory.load();
```
Comment (rcardin) [inline kafkaesque-core/src/main/java/in/rcard/kafkaesque/config/KafkaesqueConfigLoader.java:10]: No configuration of the path of the file.

File: kafkaesque-core/src/main/java/in/rcard/kafkaesque/config/KafkaesqueConsumerConfigImpl.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    private String clientId;
+    private Duration fetchMaxWait;
+    private int fetchMinSize;
+    private Duration heartbeatInterval;
+    private String isolationLevel;
+    private int maxPollRecords;
+
+    public KafkaesqueConsumerConfigImpl() {}
+
+    public String getGroupId() {
```
Comment (rcardin) [inline kafkaesque-core/src/main/java/in/rcard/kafkaesque/config/KafkaesqueConsumerConfigImpl.java:21]: Do we need the Java Bean specification? Is it possible to use the constructor instead?

File: kafkaesque-core/src/main/java/in/rcard/kafkaesque/consumer/KafkaesqueConsumer.java [hunk 7]
```diff
@@ -57,17 +60,19 @@
...
-    props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
-    props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, brokersUrl);
-    props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "true");
-    props.put(
-        ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, creationInfo.getKeyDeserializer().getClass());
-    props.put(
-        ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
-        creationInfo.getValueDeserializer().getClass());
-    final KafkaConsumer<Key, Value> consumer = new KafkaConsumer<>(props);
+    Map creationProps = Map.of(
```
Comment (rcardin) [inline kafkaesque-core/src/main/java/in/rcard/kafkaesque/consumer/KafkaesqueConsumer.java:63]: Don't use raw types in Java.

File: kafkaesque-core/src/main/java/in/rcard/kafkaesque/producer/KafkaesqueProducer.java [hunk 8]
```diff
@@ -40,14 +44,19 @@
...
   private KafkaProducer<Key, Value> createKafkaProducer(String brokerUrl) {
-    final Properties props = new Properties();
-    props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, brokerUrl);
-    props.put(ProducerConfig.ACKS_CONFIG, "all");
-    props.put(
-        ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, creationInfo.getKeySerializer().getClass());
-    props.put(
-        ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, creationInfo.getValueSerializer().getClass());
-    return new KafkaProducer<>(props);
+    Map creationProps = Map.of(
```
Comment (rcardin) [inline kafkaesque-core/src/main/java/in/rcard/kafkaesque/producer/KafkaesqueProducer.java:47]: Don't use raw types in Java.

**Review by rcardin — CHANGES_REQUESTED** [review]
Nothing major. I'm a bit concerned that we don't support YAML files (https://github.com/lightbend/config/issues/494). Anyway, it can be a first approximation of the final solution :) 

Nice work.

**Comment by rcardin** [comment]
No problem. I made some changes in the main branch. So, we will need to merge it into this branch before accepting the PR

File: README.md (hunk 4 above)
Comment (farajist) [reply README.md:121]: Yes @rcardin, as you mentioned, YAML is not supported yet by the library. I have been playing around this lib as well as [Apache Commons Configuration](https://commons.apache.org/proper/commons-configuration/index.html), the latter supports YAML but I had some issues, I will revisit it in the next iteration.

File: kafkaesque-core/src/main/java/in/rcard/kafkaesque/config/KafkaesqueConfigLoader.java (hunk 5 above)
Comment (farajist) [reply kafkaesque-core/src/main/java/in/rcard/kafkaesque/config/KafkaesqueConfigLoader.java:10]: Would the [standard behavior](https://github.com/lightbend/config#standard-behavior) be problematic ? Or is it favorable to make the path explicit :thinking:

Comment (rcardin) [reply kafkaesque-core/src/main/java/in/rcard/kafkaesque/config/KafkaesqueConfigLoader.java:10]: If there is a "standard behavior, " you must explicitly add it to the documentation. The user of `kafkaesque` should not try to understand which library we use and learn how to use that library on his own.

**Comment by rcardin** [comment]
I saw your changes. If it's ok for you, I'll clone your repo and make some changes, which we can discuss later. What's your feeling about this?

**Comment by farajist** [comment]
@rcardin sure yeah, it's totally fine !

**Comment by rcardin** [comment]
@farajist, I made some refactorings. Don't set a `public` scope for methods you don't want others to use.
By the way, there is a problem with the implementations of the `KafkaesqueProducerConfig` and `KafkaesqueConsumerConfig`. It would help if you allowed a developer not to set all the properties as you have implemented now; the application crashes.

Can you fix it? Please, try to do this using a TDD approach.

**Comment by farajist** [comment]
Thanks for the update @rcardin, I checked the refactoring that you did, amazing work ! interfaces are minimal and the `KafkaesqueConfigLoader` -> `TypesafeKafkaesqueConfigLoader` abstraction is on point. I was relying on basic testing for the config loading feature, but it would be helpful to have a sub-project for code examples to do integration testing and quickly reproduce issue, I will try and reproduce the crash scenario in order to limit the scope of configuration. Thanks !

**Comment by rcardin** [comment]
Ok, thanks. In the meanwhile, I'll continue to change the DSL part to let configure the path of the file.

**Comment by rcardin** [comment]
@farajist, do you have some time to proceed? If you don't, please, let me know. I'll finish the missing parts independently.

**Comment by farajist** [comment]
Hi @rcardin, sorry for the delay, I was busy these last days since I relocated (a different time zone), I will try to work get the changes done this weekend if possible, otherwise I will let you know, thanks and sorry for the delayed feedback :)

**Comment by farajist** [comment]
> 

As I mentioned in one of the comments

**Pull request merged**